### PR TITLE
Enable `RelionSource` to be created with no metadata

### LINKED ISSUE #519: Separate Relion-specific defaults from `ImageSource`
*WIP*

The `ImageSource` class of `aspire.source.image` contains a number of Relion-specific parameters and defaults. This causes some issues when attempting to subclass for other types of data sources. 

This issue can start as a description of parts of this class that currently rely on or set these defaults.

One of the main issues is the assumption baked into the `ImageSource` code that the data source will contain metadata related to CTF.  

- The `filter_indices` property of this class has a setter method which sets 
```python
self.set_metadata(                                                                                             
            [                                                                                                          
                "_rlnVoltage",                                                                                         
                "_rlnDefocusU",                                                                                        
                "_rlnDefocusV",                                                                                        
                "_rlnDefocusAngle",                                                                                    
                "_rlnSphericalAberration",                                                                             
                "_rlnAmplitudeContrast",                                                                               
            ],                                                                                                         
            filter_values,      
```
These correspond to similarly named properties of the `operators.filters.CTFFilter` class. If `filter_indices` is set to `None`, or if the source does not have `CTFFilter`s in its `unique_filters`, these values in metadata are all assigned to `np.nan`. Since not all sources contain CTF information, this behavior should probably either check for this, or be moved to `RelionSource`.

The `__init__` of `RelionSource` depends on this behavior of `ImageSource.filter_indices`. In this code:

```python
  filter_params, filter_indices = np.unique(                                                                     
            metadata[                                                                                                  
                [                                                                                                      
                    "_rlnVoltage",                                                                                     
                    "_rlnDefocusU",                                                                                    
                    "_rlnDefocusV",                                                                                    
                    "_rlnDefocusAngle",                                                                                
                    "_rlnSphericalAberration",                                                                         
                    "_rlnAmplitudeContrast",                                                                           
                ]                                                                                                      
            ].values,                                                                                                  
            return_inverse=True,                                                                                       
            axis=0,                                                                                                    
        )                 
```
`filter_indices` is set to a partition of the indices of `metadata` according to unique CTF parameters. But hidden in this assignment is the code in the setter above, which populates Relion-specific metadata. The sequence of events is, more or less:

```python

CTF_params = {
            "_rlnVoltage",                                                                                     
            "_rlnDefocusU",                                                                                    
            "_rlnDefocusV",                                                                                    
            "_rlnDefocusAngle",                                                                        
            "_rlnAmplitudeContrast",   
         }
if CTF_params.issubset(metadata.columns):
    filter_params, filter_indices = np.unique(metadata[list(CTF_params)])
    filters = []                                                                                                   
        for row in filter_params:                                                                                      
            filters.append(                                                                                            
                CTFFilter(                                                                                             
                    pixel_size=self.pixel_size,                                                                        
                    voltage=row[0],                                                                                    
                    defocus_u=row[1],                                                                                  
                    defocus_v=row[2],                                                                                  
                    defocus_ang=row[3] * np.pi / 180,  # degrees to radians                                            
                    Cs=row[4],                                                                                         
                    alpha=row[5],                                                                                      
                    B=B,                                                                                               
                )                                                                                                      
            )                                                                                                          
                 
        filter_values = np.zeros((len(indices), len(attribute_list)))                                              
        for i, filt in enumerate(self.unique_filters):                                                             
            filter_values[indices == i] = [                                                                        
                getattr(filt, attribute, np.nan) for attribute in attribute_list                                   
            ]
        self.set_metadata(list(CTF_params), filter_values)

```

This could probably be turned into its own method, and the extra code removed from the setter of `filter_indices`. 

- In `ImageSource.__init__()`, the `_rotations` property of `ImageSource` is populated from the `_rlnAngleRot`, `_rlnAngleTilt`, and `_rlnAnglePsi` fields, if they exist in the starfile. This should probably be moved into `RelionSource`
```python
if self.has_metadata(["_rlnAngleRot", "_rlnAngleTilt", "_rlnAnglePsi"]):                                   
                self._rotations = R.from_euler(                                                                        
                    "ZYZ",                                                                                             
                    self.get_metadata(                                                                                 
                        ["_rlnAngleRot", "_rlnAngleTilt", "_rlnAnglePsi"]                                              
                    ),                                                                                                 
                    degrees=True,                                                                                      
                )                    
```

- Several other properties of `ImageSource` explicitly return Relion STAR file parameters, which are expected to be loaded into `ImageSource.metadata`. One way around this could be to simply move these to `RelionSource`. However, some could be properties of a general `ImageSource`, so perhaps leaving the implementation to child classes would be better.
```python
  @property                                                                                                          
    def offsets(self):                                                                                                 
        return np.atleast_2d(                                                                                          
            self.get_metadata(["_rlnOriginX", "_rlnOriginY"], default_value=0.0)                                       
        )                                                                                                              
                                                                                                                       
    @offsets.setter                                                                                                    
    def offsets(self, values):                                                                                         
        return self.set_metadata(["_rlnOriginX", "_rlnOriginY"], values)                                               
                                                                                                                       
    @property                                                                                                          
    def amplitudes(self):                                                                                              
        return np.atleast_1d(self.get_metadata("_rlnAmplitude", default_value=1.0))                                    
                                                                                                                       
    @amplitudes.setter                                                                                                 
    def amplitudes(self, values):                                                                                      
        return self.set_metadata("_rlnAmplitude", values)       
```

### PR DESCRIPTION
**Pull request by chris-langfield**
Previously, a `RelionSource` would automatically look for the following CTF related parameters in the STAR file:

```
"_rlnVoltage",                                                                                                         
"_rlnDefocusU",                                                                                                        
"_rlnDefocusV",                                                                                                        
"_rlnDefocusAngle",                                                                                                    
"_rlnSphericalAberration",                                                                                             
"_rlnAmplitudeContrast", 
```

If they exist in the STAR file, `self.unique_filters` is populated with a `CTFFilter` for each unique combination of the parameters. This filter is constructed directly from the parameters above except for "_rlnDefocusAngle", which is multiplied by `np.pi / 180` to convert from degrees to radians. 

When `filter_indices` is set, behind the scenes, the `ImageSource.filter_indices` setter reads the properties of the `unique_filters` that were already set:

```python
attribute_list = (
                "voltage",
                "defocus_u",
                "defocus_v",
                "defocus_ang",
                "Cs",
                "alpha",
            )
            filter_values = np.zeros((len(indices), len(attribute_list)))
            for i, filt in enumerate(self.unique_filters):
                filter_values[indices == i] = [
                    getattr(filt, attribute, np.nan) for attribute in attribute_list
                ]
```
And then reassigns this metadata, even though it has already been read in from the STAR file (which allowed the filters to be created in the first place):

```python
self.set_metadata(
            [
                "_rlnVoltage",
                "_rlnDefocusU",
                "_rlnDefocusV",
                "_rlnDefocusAngle",
                "_rlnSphericalAberration",
                "_rlnAmplitudeContrast",
            ],
            filter_values,
        )
```

As far as I can tell, this is completely circular except for the "_rlnDefocusAngle", which was changed from degrees to radians during the construction of the filter. 

This PR

- Allows a `RelionSource` to be created from a STAR file with no metadata, i.e. one column: `_rlnImageName`. In particular, when a `CoordinateSource` is saved (#489 #520), it creates a STAR file in this format, meaning it can be loaded and saved as a `RelionSource`. 

- Removes the Relion-specific behavior of the setter method of `ImageSource.filter_indices`, which tries to populate the above parameters. Also, the table containing the data types of each Relion field has been moved from `ImageSource` to `RelionSource`. This relates to #519 where we would like to make `ImageSource` more generalized. 

- Refactors the classmethod `starfile2df` in `RelionSource` to `populate_metadata`, and adds some comments for clarity.

- Changes the default metadata storage of `_rlnDefocusAngle` from **radians** to **degrees**, to agree with how it is represented by Relion. This change is a side effect of removing the `filter_indices` setter, which would repopulate this parameter based on the `CTFFilter` attribute, which is in radians. Now it keeps the original value read from the STAR file.

### CODE REVIEW

**Comment by chris-langfield** [comment]
Squashed down to one commit. @garrettwrong I added a method in `Simulation` to populate the metadata that would have been populated by the `ImageSource.filter_indices` setter. However, it is clunky and adds lines of code. It could potentially be useful for other subclasses (e.g. if CTFFilters are added to a `CoordinateSource`). Should I move it up into `ImageSource` instead?

**Comment by janden** [comment]
The issue of storing the CTF parameters probably merits a discussion. Right now, as you point out, we store it in two places: `unique_filters` and various `metadata` fields. Ideally, there'd just be one. This is why you're seeing the copying from one place to another – to make sure that the values correspond (except for the degrees↔radians).

The root of the complication is that while we have some filters (like the CTF) that are specified by the forward model (and hence have their parameters as part of the metadata), others are given by our image processing functionality (whitening, phase flipping, downsampling, etc.). This is not so much an issue when all we're doing is processing the data within the framework, but once we want to export (write STAR files) we need to have something that fits into the CTF format again.

File: src/aspire/source/relion.py [hunk 1]
```diff
@@ -103,48 +119,100 @@ def __init__(
...
+        to the locations of particles. We read this into a Pandas DataFrame and add some of
+        our own columns for convenience.
+        """
+        if data_folder is not None:
+            if not os.path.isabs(data_folder):
+                data_folder = os.path.join(os.path.dirname(filepath), data_folder)
+        else:
+            data_folder = os.path.dirname(filepath)
+
+        # Valid Relion STAR files always have their data in the first loop of the first block.
```
Comment (garrettwrong) [inline src/aspire/source/relion.py:184]: Time will tell :)

File: src/aspire/source/relion.py [hunk 2]
```diff
@@ -103,48 +119,100 @@ def __init__(
...
+        # Note that os.path.join works as expected when the second argument is an absolute path itself
+        df["__mrc_filepath"] = df["__mrc_filename"].apply(
+            lambda filename: os.path.join(data_folder, filename)
+        )
+
+        if max_rows is None:
+            return df
+        else:
+            max_rows = min(max_rows, len(df))
+            return df.iloc[:max_rows]
```
Comment (garrettwrong) [inline src/aspire/source/relion.py:215]: Not sure if it matters here, but offhand doesn't pandas loc/iloc potentially return view? I remember fixing some bugs like that. Not saying this is wrong, but just double check this does exactly what you want.

File: src/aspire/source/relion.py [hunk 3]
```diff
@@ -162,6 +230,7 @@ def load_single_mrcs(filepath, df):
             # the code below reshapes it to (1, resolution, resolution)
             if len(arr.shape) == 2:
                 arr = arr.reshape((1,) + arr.shape)
+            # __mrc_index is the 1-based index of the particle in the stack
```
Comment (garrettwrong) [inline src/aspire/source/relion.py:233]: Good comment.

File: src/aspire/source/simulation.py [hunk 4]
```diff
@@ -96,6 +97,41 @@ def __init__(
             logger.info("Appending a NoiseAdder to generation pipeline")
             self.noise_adder = NoiseAdder(seed=self.seed, noise_filter=noise_filter)
 
+    def _populate_ctf_metadata(self, filter_indices):
+        # Since we are not reading from a starfile, we must construct
+        # metadata based on the CTF filters by hand and set the values
+        # for these columns
+        #
+        # class attributes of CTFFilter:
+        CTFFilter_attributes = (
```
Comment (garrettwrong) [inline src/aspire/source/simulation.py:115]: Not sure if these would be better as a dictionary mapping between `CTFFilter_attributes` and `CTF_params`. They look similar. At least similar enough I had to look if exactly the same...

**Review by garrettwrong — DISMISSED** [review]
I didn't realize this was waiting for me to look at, sorry.  Overall this looks good. I made a few comments.  I think only one or two things you might consider before sending to Joakim, but I didn't feel strongly enough about them to hold it up.

File: src/aspire/source/relion.py (hunk 2 above)
Comment (chris-langfield) [reply src/aspire/source/relion.py:215]: Double checked that this returns a DF when given a slice like this. Thanks!

File: src/aspire/source/simulation.py (hunk 4 above)
Comment (chris-langfield) [reply src/aspire/source/simulation.py:115]: Yeah there is a one-to-one correspondence. But the two sets of strings are used separately, so I think to do the `set_metadata` line that follows this line, I'd have to do something like `CTFDict.keys()`, which I would probably find more confusing if I were a newcomer. Unless I'm missing something

File: src/aspire/source/relion.py [hunk 5]
```diff
@@ -90,10 +106,10 @@ def __init__(
         # Get the 'mode' (data type) - TODO: There's probably a more direct way to do this.
         mode = int(mrc.header.mode)
         dtypes = {0: "int8", 1: "int16", 2: "float32", 6: "uint16"}
-        ensure(
-            mode in dtypes,
-            f"Only modes={list(dtypes.keys())} in MRC files are supported for now.",
-        )
+        assert (
```
Comment (janden) [inline src/aspire/source/relion.py:109]: Just curious about this. Why switch from `ensure` to `assert`? Are we phasing out `ensure` across the entire codebase?

**Review by janden — APPROVED** [review]

File: src/aspire/source/relion.py (hunk 5 above)
Comment (chris-langfield) [reply src/aspire/source/relion.py:109]: My impression was that we are phasing out `ensure` but maybe I misremembered? @garrettwrong 

That could go in the standards/conventions doc if so

Comment (garrettwrong) [reply src/aspire/source/relion.py:109]: Ensure has no reason to exist in python, it's just a home brew syntax for assert. Recommend to remove when encountered, converging towards using the standard library.

I think it comes from Matlab helpers, where it maybe made more sense. Fwiw.

Ensure is not a convention anyone else has, or one we should. If you prefer to remove it whole sale, go for it. :)

Comment (chris-langfield) [reply src/aspire/source/relion.py:109]: Cool, made an issue and a note on #569

**Pull request merged**